flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java

```diff
@@ -0,0 +1,486 @@
+                    fileInputStream.close();
+                }
+            } else {
+                LOG.info("Skip merging file {} because it doesn't exist", f.getName());
+            }
+        }
+
+        LOG.info("Time to merge {} s", ((System.nanoTime() - startTime) / 1000000000.0));
+    }
+
+    // Cleanup files
+    private void cleanup(List<File> inputFiles) {
+        for (File f : inputFiles) {
+            if (f.exists()) {
+                if (f.delete()) {
+                    LOG.info("Deleted temporary file {}", f.getName());
+                }
+            }
+        }
+        if (tempFolder.delete()) {
+            LOG.info("Deleted temporary folder at {}", tempFolder.getAbsolutePath());
+        }
+    }
+
+    // Get the data stream for the specified keyLength, create it if needed
+    private DataOutputStream getDataStream(int keyLength) throws IOException {
+        // Resize array if necessary
+        if (dataStreams.length <= keyLength) {
+            dataStreams = Arrays.copyOf(dataStreams, keyLength + 1);
+            dataFiles = Arrays.copyOf(dataFiles, keyLength + 1);
+        }
+
+        DataOutputStream dos = dataStreams[keyLength];
+        if (dos == null) {
+            File file = new File(tempFolder, "data" + keyLength + ".dat");
+            file.deleteOnExit();
+            dataFiles[keyLength] = file;
+
+            dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
+            dataStreams[keyLength] = dos;
+
+            // Write one byte so the zero offset is reserved
+            dos.writeByte(0);
+        }
+        return dos;
+    }
+
+    // Get the index stream for the specified keyLength, create it if needed
+    private DataOutputStream getIndexStream(int keyLength) throws IOException {
+        // Resize array if necessary
+        if (indexStreams.length <= keyLength) {
+            indexStreams = Arrays.copyOf(indexStreams, keyLength + 1);
+            indexFiles = Arrays.copyOf(indexFiles, keyLength + 1);
+            keyCounts = Arrays.copyOf(keyCounts, keyLength + 1);
+            maxOffsetLengths = Arrays.copyOf(maxOffsetLengths, keyLength + 1);
+            lastValues = Arrays.copyOf(lastValues, keyLength + 1);
+            lastValuesLength = Arrays.copyOf(lastValuesLength, keyLength + 1);
+            dataLengths = Arrays.copyOf(dataLengths, keyLength + 1);
+        }
+
+        // Get or create stream
+        DataOutputStream dos = indexStreams[keyLength];
+        if (dos == null) {
+            File file = new File(tempFolder, "temp_index" + keyLength + ".dat");
+            file.deleteOnExit();
+            indexFiles[keyLength] = file;
+
+            dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
+            indexStreams[keyLength] = dos;
+
+            dataLengths[keyLength]++;
+        }
+        return dos;
+    }
+
+    private int getNumKeyCount() {
+        int res = 0;
```

Comment:
```
return Arrays.stream(keyCounts).filter(c -> c != 0).count;
```

Reply:
In performance critical path, we don't use java stream.